instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/FixedSizeCache.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.api;
+
+/**
+ * This is a fixed size cache that only has one operation <code>computeIfAbsent</code>, that is used
+ * to retrieve, or store and compute the cached value.
+ *
+ * <p>If there is a hash collision, the cache uses double hashing two more times to try to find a
+ * match or an unused slot.
+ *
+ * <p>The cache is thread safe, and assumes that the <code>Creator</code> passed into <code>
+ * computeIfAbsent</code> is idempotent or otherwise you might not get back the value you expect
+ * from a cache lookup.
+ *
+ * @param <K> key type
+ * @param <V> value type
+ */
+public final class FixedSizeCache<K, V> {
+
+  static final int MAXIMUM_CAPACITY = 1 << 30;
+
+  private static final class Node<K, V> {
+    private final K key;
+    private final V value;
+
+    private Node(K key, V value) {
+      this.key = key;
+      this.value = value;
+    }
+  }
+
+  private final int mask;
+  // This is a cache, so there is no need for volatile, atomics or synchronized.
+  // All race conditions here are benign since you always read or write a full
+  // Node that can not be modified, and eventually other threads will see it or
+  // write the same information at that position, or other information in the
+  // case of a collision.
+  private final Node<K, V>[] elements;
+
+  /**
+   * Creates a <code>FixedSizeCache</code> that can hold up to <code>capacity</code> elements, if
+   * the key hash function has perfect spread.
+   *
+   * @param capacity the maximum number of elements that the cache can hold
+   */
+  public FixedSizeCache(int capacity) {
+    if (capacity <= 0) {
+      throw new IllegalArgumentException("Cache capacity must be > 0");
+    }
+    if (capacity > MAXIMUM_CAPACITY) {
+      capacity = MAXIMUM_CAPACITY;
+    }
+    // compute a power of two size for the given capacity
+    int n = -1 >>> Integer.numberOfLeadingZeros(capacity - 1);
+    n = (n < 0) ? 1 : (n >= MAXIMUM_CAPACITY) ? MAXIMUM_CAPACITY : n + 1;
```

Comment:
This line needs a comment since it's pretty hard to read. Though I'd generally recommend just copying from the leading expert on caches

https://github.com/ben-manes/caffeine/blob/37f6ad303eb4474cd8a644551d528dfda37c5bfc/caffeine/src/main/java/com/github/benmanes/caffeine/SingleConsumerQueue.java#L136

Reply:
Hey @bantonsson, you should have just copied this ;)